### ISSUE TITLE
QA training: segmentation fault with BERT large

### ISSUE THREAD

**Issue by ckobus** [deepset-ai/FARM#178]
Hi all,

I have a question with respect to QA training with BERT large on squad 2.0.
I tried to reproduce the QA training you mention in the blog post https://medium.com/deepset-ai/modern-question-answering-systems-explained-4d0913744097, using the BERT large, with batch_size=32 and max_seq_len=384. But I have a crash at the very beginning of the training (Segmentation fault).
When reducing the batch_size to 16, it seems to work (training still ongoing).
I think I am using the same machine as you mention, it 's a p3.8xlarge instance on AWS.

Did you use some trick to limit the GPU memory usage or did you use another kind of instance?

Thank for your help!!

**Comment by Timoeller** [comment]
Hey Catherine,

the experiments, parameters and performances that were reported in Figure 9 + 10 are related to the Bert base version, not the Bert large version. This one is of courses larger and needs more memory. 
Though you already found the solution by lowering the batch size. 

If this large model should not converge you could try using gradient_accumulation over multiple batch updates. You can set this value with parameter "grad_acc_steps" in initialize_optimizer() function and have to pass it to the Trainer class.

Hope that helps?
Would be nice if you can report your scores with the large model once it is finished, as reference for the community.

**Comment by ckobus** [comment]
Thanks for the quick answer and the tip about gradient_accumulation; I will try to use it...

But it seems the QA model you provide with the Haystack tutorial comes from a Bert large, no?

In the notebook https://github.com/deepset-ai/haystack/blob/master/tutorials/Tutorial1_Basic_QA_Pipeline.ipynb, there is a link to the model url="https://s3.eu-central-1.amazonaws.com/deepset.ai-farm-models/0.3.0/bert-english-qa-large.tar.gz" 
In the tar.gz, the are different config json files and in the language model related one, it seems it is a bert large. 
Actually, by looking at the performance_and_params.Txt file, a batch size of 20 was used.
At the end, there are results with squad script; are there on the dev set? if yes, it seems using bert large, you gain around 9 points both in EM and F1, compared to Bert base.

**Comment by Timoeller** [comment]
No worries, always happy to help.

> [...]  the QA model you provide with the Haystack tutorial comes from a Bert large, no?

Exactly, this model is a large one and has quite some performance improvements over the base version... at the cost of training and inference speed. The performance is on the dev set, the SQuAD test set is actually not public, you need to upload your model. This uploaded model is evaluated on own SQuAD servers on the private test set.  

> Actually, by looking at the performance_and_params.Txt file, a batch size of 20 was used.

If we used a batch size of 20, you should also be able to pack that onto your 4xV100 machine I hope.
Btw. you can have a look at the memory footprint of your GPU with "nvidia-smi -l 2" in the cmd line on the server and adjust the batch size to completely fill up memory. To get the most out of your GPU you actually need to consider much more, but for quick tests varying the batch_size only should be fine.

Did you succeed in training your own model? How did it perform? If you did not use a Bert large model ending with "whole-word-masking", your results should be 1-3%points less on SQuAD dev.

**Comment by ckobus** [comment]
The training is almost finished but actually, it seems having issues to converge.
the best result I got so far on the dev is after 8000 iterations:

12/19/2019 13:06:49 - INFO - farm.eval -   loss: 0.8275824179122975
12/19/2019 13:06:49 - INFO - farm.eval -   task_name: question_answering
12/19/2019 13:06:49 - INFO - farm.eval -   EM: 0.7626547629074371
12/19/2019 13:06:49 - INFO - farm.eval -   f1: 0.801419766519876
12/19/2019 13:06:49 - INFO - farm.eval -   report:

Since then, the loss is increasing on the dev.

I use Bert-large-cased. In general and in other downstream tasks, Bert with "whole-word-masking" is better?

**Comment by ckobus** [comment]
Another question, which value would you recommend for the parameter grad_acc_steps?

**Comment by Timoeller** [comment]
Yes, models produced with whole word masking should perform consistently better across tasks (though only slightly).
Since you already have the link to a Bert large WWM fine-tuned on squad with batch size 20, I would recommend to try something new :)
What about setting gradient accumulation to 2 with batch size 20, resulting in an effective batch size of 40. Maybe you can beat the performance mentioned in the txt file you mentioned?
